ansible_runner/streaming.py

```diff
@@ -0,0 +1,227 @@
+            except ConfigurationError:
+                settings = {}
+        self.config = MockConfig(settings)
+
+        artifact_dir = kwargs.get('artifact_dir')
+        self.artifact_dir = os.path.abspath(
+            artifact_dir or os.path.join(self.private_data_dir, 'artifacts'))
+
+        self.status_handler = status_handler
+        self.event_handler = event_handler
+        self.artifacts_handler = artifacts_handler
+
+        self.cancel_callback = cancel_callback  # FIXME: unused
+        self.finished_callback = finished_callback
+
+        self.status = "unstarted"
+        self.rc = None
+
+    def status_callback(self, status_data):
+        self.status = status_data['status']
+        if self.status == 'starting':
+            self.config.command = status_data.get('command')
+            self.config.env = status_data.get('env')
+            self.config.cwd = status_data.get('cwd')
+
+        for plugin in ansible_runner.plugins:
+            ansible_runner.plugins[plugin].status_handler(self.config, status_data)
+        if self.status_handler is not None:
+            self.status_handler(status_data, runner_config=self.config)
+
+    def event_callback(self, event_data):
+        full_filename = os.path.join(self.artifact_dir,
+                                     'job_events',
+                                     '{}-{}.json'.format(event_data['counter'],
+                                                         event_data['uuid']))
+        if 'stdout' in event_data:
+            print(event_data['stdout'])
+
+        if self.event_handler is not None:
+            should_write = self.event_handler(event_data)
+        else:
+            should_write = True
+        for plugin in ansible_runner.plugins:
+            ansible_runner.plugins[plugin].event_handler(self.config, event_data)
+        if should_write:
+            with codecs.open(full_filename, 'w', encoding='utf-8') as write_file:
+                os.chmod(full_filename, stat.S_IRUSR | stat.S_IWUSR)
+                json.dump(event_data, write_file)
+
+    def artifacts_callback(self, artifacts_data):
+        buf = io.BytesIO(artifacts_data)
+        with zipfile.ZipFile(buf, 'r') as archive:
+            archive.extractall(path=self.config.artifact_dir)
+
+        if self.artifacts_handler is not None:
+            self.artifacts_handler(self.config.artifact_dir)
+
+    def run(self):
+        job_events_path = os.path.join(self.artifact_dir, 'job_events')
+        if not os.path.exists(job_events_path):
+            os.makedirs(job_events_path, 0o700, exist_ok=True)
+
+        while True:
+            line = self._input.readline()
+            data = json.loads(line)
+
+            if 'status' in data:
+                self.status_callback(data)
+            elif 'zipfile' in data:
+                self.artifacts_callback(self._input.read(data['zipfile']))
+            elif 'eof' in data:
+                break
```

Comment:
AWX will still want the event_callback to be invoked for the EOF event.

Reply:
I don't believe that that is the way it works.  I think the unconditional call of `finished_callback` below correctly parallels how it currently works.

Reply:
It's hard to talk precisely about this since this change isn't the full picture, but yes, it's true that someone (AWX) using this pragmatically could pass a `finished_callback` that re-implements the construction of the EOF dict at:

https://github.com/ansible/ansible-runner/blob/e6020b7d33627a5d6656dca64d7ae09dd95cc75c/ansible_runner/utils.py#L325

The point I was trying to make is that AWX makes decisions upon detection of that EOF event, concretely:

https://github.com/ansible/awx/blob/0c7bfa543b9e1b5637b69c97696e9ea145a0b98f/awx/main/dispatch/worker/callback.py#L158

To have that continue to work, it seems pretty cheap to me to add

```python
self.event_callback(data)
```

before `break`. It's not that you couldn't resolve that in the `finished_callback`... or you could do something more dramatic, like move logic currently in the callback receiver to `tasks.py` where it inputs the callback methods. But there's a lot invested in this AWX EOF logic, and it would be good to leave that untouched as much as possible.

Reply:
That's not what I'm saying at all, though the issue is confounded by our inconsistent use of the terms 'callback' and 'handler'.

Reply:
Ok, so, here's the deal.  We have _two_ different things that are an EOF.  We have `{'event': 'EOF'}` sent from `utils.OutputEventFilter` unconditionally to `Runner.event_callback`.  None of my work here intercepts or does anything different with this event.  Secondly, we have `{'eof': True}` as issued by `Worker.finished_callback`.  This is only intended to be a message used to break out of the loop in `Process.run`.